src/compiler.rkt

```diff
@@ -219,3 +194,168 @@
 ;; Like `compile-progs`, but a single prog.
 (define (compile-prog expr ctx)
   (compose first (compile-progs (list expr) ctx)))
+
+(define (backward-pass ivec varc vregs vprecs vstart-precs)
+  (vector-fill! vprecs 0)
+  (for ([instr (in-vector ivec (- (vector-length ivec) 1) -1 -1)]
+        [n (in-range (- (vector-length vregs) 1) -1 -1)])
+
+    (define op (car instr))
+    (define tail-registers (rest instr))
+    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers))
+    (define output (vector-ref vregs n))
+
+    (define exps-from-above (vector-ref vprecs n))
+    (define final-parent-precision (min (*max-mpfr-prec*) (+ exps-from-above (vector-ref vstart-precs n))))
+    (when (equal? final-parent-precision (*max-mpfr-prec*))
+      (*sampling-iteration* (*max-sampling-iterations*)))
+    (vector-set! vprecs n final-parent-precision)
+
+    (define new-exponents 0)
+    (cond
+      [(equal? op ival-add)
+       (define xlo (ival-lo (first srcs)))
+       (define xlo-exp (true-exponent xlo))
+       (define xlo-sgn (bigfloat-signbit xlo))
+       (define ylo (ival-lo (second srcs)))
+       (define ylo-exp (true-exponent ylo))
+       (define ylo-sgn (bigfloat-signbit ylo))
+       (define outlo-exp (true-exponent (ival-lo output)))
+       ; consider only lower bound
+       (set! new-exponents
+             (max 0
+                  (match (and (not (equal? xlo-sgn ylo-sgn))
+                              (>= 1 (abs (- xlo-exp ylo-exp))))
+                    [#f (- (max xlo-exp ylo-exp) outlo-exp)]
+                    [#t (+ (get-slack) (- (max xlo-exp ylo-exp) outlo-exp))])))]
+      
+      [(equal? op ival-sub)
```

Comment:
This is the same as `ival-add` except that the same-sign test is the same, right?

Reply:
Yeah, they are almost the same
I just treat add as two operations:
`outlo = xlo + ylo`, `outhi = xhi + yhi`
when sub:
`outlo = xlo - yhi`, `outhi = xhi - ylo`
And basically I take the max of condition number of these two operations.
That's why the formulas are sightly different - `ylo` replaces with `yhi`
This affect sign test and exponent calculation by itself

Reply:
Ok.